### PR TITLE
Transducer v4

### PR DESCRIPTION
**Pull request by b-flo**
**Note: PR re-created in a tentative to fix codecov, refer to #2082 for discussions**

This PR add/modify a bunch of things:

To do / in progress:

- Default beam search:
  - [x] Optimization techniques
    - [x] prediction net caching
- N-Step Constrained beam search (modified version of: https://arxiv.org/pdf/2002.03577.pdf): 
  - [x] RNN-T
  - [x] RNN-T w/ att.
  - [x] T-T
  - [x] VIVOS decode config
  - [x] Voxforge decode config
  - [x] VIVOS results
  - [x] Voxforge results
  - [x] Optimization techniques
    - [x] prediction net caching
  - [ ] Code optimization
- Time Synchronous Decoding (https://ieeexplore.ieee.org/document/9053040): 
  - [x] RNN-T
  - [x] RNN-T w/ att.
  - [x] T-T
  - [x] VIVOS decode config
  - [x] Voxforge decode config
  - [x] VIVOS results
  - [x] Voxforge results
  - [x] Optimization techniques
    - [x] prediction net caching
    - [x] prediction net bashing
    - [x] joint net caching
  - [ ] Code optimization
- Alignment-Length Synchronous Decoding (https://ieeexplore.ieee.org/document/9053040):
  - [x] RNN-T
  - [x] RNN-T w/ att.
  - [x] T-T
  - [x] VIVOS decode config
  - [x] Voxforge decode config
  - [x] VIVOS results
  - [x] Voxforge results
  - [x] ! Fix bug final hypothesis empty with transformer (Edit: lazy patch for now, I'll investigate later)
  - [x] Optimization techniques
    - [x] prediction net caching
    - [x] prediction net bashing
    - [x] joint net caching
  - [ ] Code optimization
- Transformer-Transducer:
  - [x] Customizable architecture
  - [x] TDNN-BN-ReLU blocks for encoder part
  - [x] Conformer blocks for encoder part
  - [x] CausalConv1d for decoder part
- General:
  - [x] CUDA-Warp RNN-T support (https://github.com/1ytic/warp-rnnt) 
  - [x] Beam search interface (first version)
  - [x] Beam search tests (I'll add more tests in another PR I think)
  - [x] Benchmark: decoding speed (edit: only rough benchmark for now, complete will be done after optimizations)
  - [x] Fix mtl_mode (#2162)
  - [x] Fix attention weights saving condition + PlotAttentionReport selection for unified transducer architecture
  - [x] Fix CER/WER reporting CPU-GPU bug (#2232)
  - [x] Fix CER/WER reporting multi GPUs bug (#2084)
  - [x] Documentation
      - [x] T-T w/ customizable architecure
      - [x] New features in main README
- Recipe
  - [x] Voxforge: remove transducer run script + transfer learning scripts for voxforge (too complicated to maintain and documentation is now available for finetuning)
  - [x] Speed-up transducer-related tests execution time #2459

Note/Important: Modifications related to customizable architecture should be discussed. This version is only intended to start the discussion: it works correctly however it's not user-friendly. This feature would be better suited with espnet2 but I think it should be accessible in espnet1 for now as it is necessary to reproduce some (if not most) transformer-transducer related papers.

To be added later: prefix tree when dealing with BPE.

~P.S: I'm now focused on porting transducer to espnet2 so I'll probably keep remaining work on standby.~

### CODE REVIEW

**Comment by kamo-naoyuki** [comment]
Circle CI doesn't exist...

**Comment by b-flo** [comment]
I'm not too familiar with Circle CI but I'll try to find a fix. At least I now know the exact reason, I was a bit confused by the fact codecov was updated every time anyways.

**Comment by ShigekiKarita** [comment]
I'm afraid that we ran out of circle ci
image
https://app.circleci.com/settings/plan/github/espnet/overview?return-to=https%3A%2F%2Fapp.circleci.com%2Fpipelines%2Fgithub%2Fespnet

**Comment by b-flo** [comment]
@ShigekiKarita I sped-up tests related to RNN-T & SA-T but I don't think they can run under 3s, notably the later. Each file "encapsulate" several test (e.g: conformer, transformer, tdnn, custom architecture, all search algorithms, etc, are tested in `test_e2e_asr_sa_transducer.py`).
Is that an issue? Should I divide each file into several files?

**Comment by ShigekiKarita** [comment]
I also found under 3s is too difficult. I think smaller files are better for readability and refactoring.

**Comment by b-flo** [comment]
@ShigekiKarita I'll further refine and split files but I don't think there will be a major difference in execution time. Below some benchmark between commit #98dof55 (when I added all new test) and last commit:

Execution time:

| test | before (s.) | after (s.) |
| - | - | - |
| test_e2e_asr_transducer.py | 13.02 | 12.49 |
| test_e2e_asr_sa_transducer.py | 43.5 | 8.64 |

Top 5 slowest test:

```
2.58s call     test/test_e2e_asr_sa_transducer.py::test_sa_transducer_parallel                                                                                            
0.61s call     test/test_e2e_asr_sa_transducer.py::test_calculate_plot_attention                                                                                          
0.13s call     test/test_e2e_asr_sa_transducer.py::test_sa_transducer_trainable_and_decodable[train_dic0-recog_dic0]
0.10s call     test/test_e2e_asr_transducer.py::test_pytorch_transducer_trainable_and_decodable[train_dic36-recog_dic36]                                                  
0.09s call     test/test_e2e_asr_transducer.py::test_pytorch_transducer_trainable_and_decodable[train_dic37-recog_dic37]
```

Coverage related to transducer : 

| before (%) | after (%) |
| - | - |
| 92.5 | 93.875 |

**Comment by b-flo** [comment]
@kamo-naoyuki I added support for 1ytic's CUDA `warp_rnnt` but I'm not sure how installation/usage should be handled. From what I tested, it seemed to work only with torch=1.1.0 and cuda=10.0. I ended-up with multiple errors during import in other cases.

Right now, I only install `warp_rnnt` if the versions are matched and espnet is not installed in cpu mode. As `warp_transducer` is used by default, I consider it optional in `check install.py` but notify user if not installed. I also raise an error during training if the lib is not installed and `trans-type: 'warp-rnnt'` is set.

P.S: Dunno if it helps but CI test fail when installing `black`, I had the same issue today and I uninstalled `typing` before re-installing.

**Comment by kamo-naoyuki** [comment]
Thank, about typing, please ignore. I'll check.

> I ended-up with multiple errors during import in other cases.

Could you share me the error?

**Comment by b-flo** [comment]
Sure!

I don't have all the encountered errors but the following was the most common (CUDA_VERSION=10.1, TH_VERSION=1.1.4):

```
  File "/home/fboye/espnet/tools/venv/lib/python3.7/site-packages/warp_rnnt-0.4.0-py3.7-linux-x86_64.egg/warp_rnnt/__init__.py", line 2, in <module>
    import warp_rnnt._C as core
ImportError: /home/fboye/espnet/tools/venv/lib/python3.7/site-packages/warp_rnnt-0.4.0-py3.7-linux-x86_64.egg/warp_rnnt/_C.cpython-37m-x86_64-linux-gnu.so: undefined symbol: cudaSetupArgument
```

P.S: forgot to add in previous message, the issue with `typing` occurs when importing a backport instead of the stdlib module in python 3.7. Thus, `typing` should be ignored if python >= 3.7

**Comment by b-flo** [comment]
In fact it seems to be less restrictive, requirement is only CUDA_VERSION=10.0. Dunno why it failed with different TH_VERSION when I tested yesterday.

**Comment by ShigekiKarita** [comment]
> @ShigekiKarita I'll further refine and split files but I don't think there will be a major difference in execution time. Below some local benchmark between commit #98dof55 (when I added all new test) and last commit:

Thanks for reducing testing time. It looks good enough to me.

Now you have done everything except for code optimizations and bug fix (#2084)? If so, I think it is high time to merge this PR before working on them.

File: test/net_definition/decoder_causalconv_transformer.json
Comment (ShigekiKarita) [inline test/net_definition/decoder_causalconv_transformer.json:1]: Where are these json files used?

**Review by ShigekiKarita — APPROVED** [review]

File: test/net_definition/decoder_causalconv_transformer.json
Comment (b-flo) [reply test/net_definition/decoder_causalconv_transformer.json]: They were used when custom network definition was defined outside training config, thanks!

**Comment by b-flo** [comment]
> Now you have done everything except for code optimizations and bug fix (#2084)? If so, I think it is high time to merge this PR before working on them.

Code optimization will be done in another PR. I couldn't find a proper "solution" at python level so I think I'll do that at C++ level as @takaaki-hori suggested.
Sorry, I forgot about the multi-gpu bug, I'll take a look again later this day!

File: egs/mini_an4/asr1/conf/train_conformer_transducer.yaml
Comment (sw005320) [inline egs/mini_an4/asr1/conf/train_conformer_transducer.yaml:38]: Is there a way to force `mtlalpha: 1.0` internally in the transducer mode or whatever, and remove this config?
This will help to avoid some confusions.

File: egs/vivos/asr1/RESULTS.md
Comment (sw005320) [inline egs/vivos/asr1/RESULTS.md:1]: Can you upload the model?
This will help people who only focus on decoding configurations.

File: egs/vivos/asr1/conf/tuning/transducer/train_transformer-rnn_transducer.yaml
Comment (sw005320) [inline egs/vivos/asr1/conf/tuning/transducer/train_transformer-rnn_transducer.yaml:42]: ditto

File: egs/voxforge/asr1/RESULTS.md
Comment (sw005320) [inline egs/voxforge/asr1/RESULTS.md:127]: ditto.
Uploading models is very useful.

**Review by sw005320 — APPROVED** [review]
LGTM.
My comments are not mandatory.
We can do it in the later PR or leave it as it is.

- I think you can keep `mtlalpha: 1.0` in the config by following a similar manner to CTC, but I just feel that it is a little bit confusing.
- Is there some consistency test that the greedy search method and each beam search method with beamsize 1 are equivalents. That is helpful to easily detect that the beam search is not broken in this unique setup at least.

File: egs/mini_an4/asr1/conf/train_conformer_transducer.yaml
Comment (b-flo) [reply egs/mini_an4/asr1/conf/train_conformer_transducer.yaml]: Thanks, forgot to change that part, there was only an assert to remove. Because of some recent changes, we don't need to force `mtlalpha` to 1 for transducer anymore.

**Comment by b-flo** [comment]
> Is there some consistency test that the greedy search method and each beam search method with beamsize 1 are equivalents. That is helpful to easily detect that the beam search is not broken in this unique setup at least.

I'll add more tests for beam search algorithms in another PR if that's okay with you!

Also, I'll upload models tomorrow or the day after as I need to re-run most experiments because of some recent changes I didn't take into account...

**Comment by sw005320** [comment]
> > Is there some consistency test that the greedy search method and each beam search method with beamsize 1 are equivalents. That is helpful to easily detect that the beam search is not broken in this unique setup at least.
> 
> I'll add more tests for beam search algorithms in another PR if that's okay with you!
> 
> Also, I'll upload models tomorrow or the day after as I need to re-run most experiments because of some recent changes I didn't take into account...

That is great, but this is not mandatory.
I'm actually OK to merge this PR in the current stage, and we could do the rest of the things in the later PR.

**Comment by b-flo** [comment]
I'll at least re-run experiments just to be sure I didn't mess something as I changed a lot of parameters over time. I'll add models at the same time. Sorry, it won't take long, max one day!

**Comment by b-flo** [comment]
I update VIVOS results and models. I didn't know which ones I should add so, well, I added all of them. I'll update Voxforge results and models in another PR

**Comment by b-flo** [comment]
@kamo-naoyuki I guess we can merge after your approval! I'll open another PR for transducer port to espnet2 with minimal implementation as draft next to discuss what we can bring from this PR.

**Comment by ShigekiKarita** [comment]
Hmm, there were strange Circle CI failures. Can you fetch and merge master? There are lots of changes in testing recently.

**Comment by b-flo** [comment]
@ShigekiKarita Hum I'm quite confused in this one. rnn-t decoding fails on:

`Too long with no output (exceeded 10m0s): context deadline exceeded`

Also, it seems the decoding for this test (`ASR, backend=pytorch, model=rnnt`) is done with LM unlike before.

If I try to debug locally, I end up with an error at first integration test (`ASR, backend=pytorch lm=RNNLM`):

`Data file '/home/fboye/espnet/egs/mini_an4/asr1/.coverage' doesn't seem to be a coverage data file: Couldn't use data file '/home/fboye/espnet/egs/mini_an4/asr1/.coverage': no such table: coverage_schema`

I'm not sure what's the issue here...

**Comment by b-flo** [comment]
> If I try to debug locally, I end up with an error at first integration test (ASR, backend=pytorch lm=RNNLM):
>
> `Data file '/home/fboye/espnet/egs/mini_an4/asr1/.coverage' doesn't seem to be a coverage data file: Couldn't use data file '/home/fboye/espnet/egs/mini_an4/asr1/.coverage': no such table: coverage_schema`
>
> I'm not sure what's the issue here...

Nevermind, I can just remove coverage option for local test, sorry.

> Hum I'm quite confused with this one: in some cases, first integration test for transducer fails during decoding with:
> 
> `Too long with no output (exceeded 10m0s): context deadline exceeded`

Dunno why it happened as I can't reproduce the error locally but I suppose it was due to the break loop condition in default beam search and previous beam size value. It's all good now.

Only remaining failing test is due to a HTTP 429.

**Review by kamo-naoyuki — APPROVED** [review]
LGTM. Thanks for really hard and great works!

If you don't mind, please open PRs one by one at the next time, I mean, a PR should be smallest to represent the proposal. Otherwise, it's hard to detect all changes for the other people. 
(Of course, there are some exceptions, e.g some drastic changing.)

Maybe, we'll need some discussions for the implementation for espnet2, so please start a PR with small status. I'd like to help you as possible I can.

File: espnet/nets/pytorch_backend/transducer/loss.py
Comment (kamo-naoyuki) [inline espnet/nets/pytorch_backend/transducer/loss.py:39]: Oh, it's not allowed to use sys.exit.
Please raise the error as it is.

File: espnet/nets/pytorch_backend/transducer/causal_conv1d.py
Comment (kamo-naoyuki) [inline espnet/nets/pytorch_backend/transducer/causal_conv1d.py:24]: This is just a comment.
Please avoid python2 syntax in espnet2.
e.g.

```
# bad
        super(CausalConv1d, self).__init__()
# good
        super().__init__()
```

You must write type annotations instead of type information in docstring in epsnet2. This is mandatory.

**Comment by b-flo** [comment]
> If you don't mind, please open PRs one by one at the next time, I mean, a PR should be smallest to represent the proposal. > Otherwise, it's hard to detect all changes for the other people.
> (Of course, there are some exceptions, e.g some drastic changing.)

Yes, sorry, I totally agree. This PR got out of my hands, I should have only added customizable architecture and some refactoring at the beginning but well, features been piled up..

> Maybe, we'll need some discussions for the implementation for espnet2, so please start a PR with small status. I'd like to help you as possible I can.

Sure, I'll add transducer the way you intended (from the placeholders I found) and we can start the discussion from here!

File: espnet/nets/pytorch_backend/transducer/loss.py
Comment (b-flo) [reply espnet/nets/pytorch_backend/transducer/loss.py:39]: Oh sorry, I wanted to exit without stopping the process (`exit(1)`) because warp-rnnt install is optional. What would be the best option in that case?

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/transducer/loss.py:39]: Or, raise RuntimeError with the message.

File: espnet/nets/pytorch_backend/transducer/causal_conv1d.py
Comment (b-flo) [reply espnet/nets/pytorch_backend/transducer/causal_conv1d.py:24]: OK! However, as far as I know, this syntax is used almost everywhere in espnet1? Should I change that for all classes related to this PR?

File: espnet/nets/pytorch_backend/transducer/loss.py
Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/transducer/loss.py:39]: Maybe, you misunderstood. sys.exit(1) is almost equivalent to raise SystemExit, so it will stop the process.

I'm still not sure your intension. Do you mean it should be fallback to warp-transducer?

Comment (b-flo) [reply espnet/nets/pytorch_backend/transducer/loss.py:39]: Oh, sorry misunderstood what part it was, will change that!

File: espnet/nets/pytorch_backend/transducer/causal_conv1d.py
Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/transducer/causal_conv1d.py:24]: No no, this is just a comment about espnet2. Sorry for confusing.

Ideally saying, espnet1 should be refactored everywhere, not only about this point, but I gave up to refactor espnet1 once because espnet1 has too much problems. This is main reason why I decided to create espnet2.

**Comment by kamo-naoyuki** [comment]
Sorry, I didn't follow all discussions. warp-rnnt is faster than warp-transducer? The training result is same? Is it worth to use?

File: espnet/nets/pytorch_backend/transducer/rnn_decoder.py
Comment (kamo-naoyuki) [inline espnet/nets/pytorch_backend/transducer/rnn_decoder.py:90]: Did you follow #2492 #2169 ?

`to_device(self, tensor)` can't be used for pytorch>=1.5.
We must solve this problem for training mode at least, so It's not better to use to_device in `forward()`. It's okay to use in decode function, I think.

Could you fix it at the next PR?

**Comment by b-flo** [comment]
> Sorry, I didn't follow all discussions. warp-rnnt is faster than warp-transducer?

From my (few) experiments, I didn't observe any changes in training time but it should be. As @sw005320 said I have to do some testing on edge cases but I don't have too much time for that right now. I hope someone can take that task or I'll wait for users feedback.
Also, after espnet2 porting, I plan to work on our own torch lib for training (w/ support monotonic rnnt and some optimizations) + decoding transducer.

> The training result is same?

Yes!

> Is it worth to use?

I guess it's worth the support as a few people asked for it!

**Comment by kamo-naoyuki** [comment]
> > Is it worth to use?
> 
> I guess it's worth the support as a few people asked for it!

Okay, thanks!

File: espnet/nets/pytorch_backend/transducer/rnn_decoder.py
Comment (b-flo) [reply espnet/nets/pytorch_backend/transducer/rnn_decoder.py:90]: I didn't, sorry! I'll fix that in another PR!

File: espnet/nets/pytorch_backend/transducer/loss.py
Comment (b-flo) [reply espnet/nets/pytorch_backend/transducer/loss.py:39]: Oh my, I understand why I did that: I was tired and lazy when adding test for warp-rnnt and used a sys.exit here to avoid a failure with pytest...

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/transducer/loss.py:39]: Hmm?

Comment (b-flo) [reply espnet/nets/pytorch_backend/transducer/loss.py]: Don't mind, sorry, I changed that!

**Comment by kamo-naoyuki** [comment]
Okay, thanks, then I'll merge this PR after CIs are finsihed.

**Comment by sw005320** [comment]
Many congrats, @b-flo!
This is one of the biggest PRs, and I'm very excited that it was merged finally!

**Comment by albertz** [comment]
This is a nice and clean implementation. It's nice to have the comparison of different beam search algorithms.

You mentioned that you wrote and submitted a paper about the N-Step Constrained (NSC) beam search? What paper is that?

I wonder, your initial hyp is `Hypothesis(score=0.0, yseq=[self.blank], dec_state=dec_state)`. Does that mean that you feed `self.blank` as the initial LM token? Shouldn't you use sth like a begin-of-sentence (BOS) token? (Which might be the same as the EOS token.) Or is blank just the same token id?

I also wonder, shouldn't you also add the EOS score of the LM at the end?

**Comment by b-flo** [comment]
Hi,

> This is a nice and clean implementation. It's nice to have the comparison of different beam search algorithms.

Thank you!

> You mentioned that you wrote and submitted a paper about the N-Step Constrained (NSC) beam search? What paper is that?

Most of the related work is done but no paper is submitted yet for the following reasons: 
 - I planned to submit a paper focused only on beam search decoding and expansion problem but I put it on standby because we (ESPnet members) decided recently to accelerate works related to Transducer and streaming in ESPnet. In the end, I think it will be included in one of the paper projects rather than standalone.
 - There is an important part related to the algorithm (and the modification of OSC) I couldn't fully explain nor validate which bother me greatly...

Btw, note that NSC beam search in ESPnet works as intended but the algorithm is not in its complete form because of some things not publicly explained. For example, the current condition for N=1 may seems weird for most people (incomplete hypothesis, in regards to blank prob, are carried to the next step).

> I wonder, your initial hyp is Hypothesis(score=0.0, yseq=[self.blank], dec_state=dec_state). Does that mean that you feed self.blank as the initial LM token? Shouldn't you use sth like a begin-of-sentence (BOS) token? (Which might be the same as the EOS token.) Or is blank just the same token id?

Yes, blank symbol is used as BOS/EOS compared to the MTL approach which use blank + BOS/EOS for CTC + att.

> I also wonder, shouldn't you also add the EOS score of the LM at the end?

That's a good question. I remember adding EOS score for multi-level word LM some time ago but I can't remember/find my conclusion on why it was not added in ESPnet. I'll revisit that part shortly, thanks for the reminder.

**Pull request merged**
